nautobot/docs/development/natural-keys.md

```diff
@@ -0,0 +1,135 @@
+
+- One or more `UniqueConstraint` declarations
+- Any `unique_together` declaration
+- Any field (other than `id`) that is set as `unique=True`.
+
+There are a few special cases that will need special handling.
+
+### Self-Referential Natural Keys
+
+An example of this can be seen with the `Location` model, where a given instance is only uniquely identified by its name **in combination with its parent**, which is another `Location`. Nautobot's default implementation would fall into an infinite recursion when trying to identify the Location's natural key fields, since they would be identified as  `("name", "parent__name", "parent__parent__name", "parent__parent__parent__name", ...)`.
+
+In a case like this, Nautobot is able to support _variadic_ natural keys, where the number and listing of natural keys may vary depending on the data of a given instance. To make this work, you will need to override two natural key related APIs on your model, `natural_key_field_lookups` and `natural_key_args_to_kwargs`, as follows:
+
+```python
+class Location(TreeModel):
+
+    class Meta:
+        unique_together = [["parent", "name"]]
+
+    @classproperty
+    def natural_key_field_lookups(cls):
+        """
+        Due to the recursive nature of Location's natural key, we need a custom implementation of this property.
+
+        This returns a set of natural key lookups based on the current maximum depth of the Location tree.
+        For example if the tree is 2 layers deep, it will return ["name", "parent__name", "parent__parent__name"].
+
+        Without this custom implementation, the generic `natural_key_field_lookups` would recurse infinitely.
+        """
+        lookups = []
+        name = "name"
+        for _ in range(cls.objects.max_tree_depth() + 1):
+            lookups.append(name)
+            name = f"parent__{name}"
+        return lookups
+
+    @classmethod
+    def natural_key_args_to_kwargs(cls, args):
+        """Handle the possibility that more recursive "parent" lookups were specified than we initially expected."""
+        args = list(args)
+        natural_key_field_lookups = list(cls.natural_key_field_lookups)
+        while len(args) < len(natural_key_field_lookups):
+            args.append(None)
+        while len(args) > len(natural_key_field_lookups):
+            natural_key_field_lookups.append(f"parent__{natural_key_field_lookups[-1]}")
+        return dict(zip(natural_key_field_lookups, args))
+```
+
+### Natural Keys Referencing a Different Self-Referential Model
+
+Relatedly, if you have a model whose natural keys include a ForeignKey to a model like `Location` with self-referential and variadic natural keys, for the related model to be handled properly, you must always ensure that the related field is the **last** such field in your model's uniqueness constraint or `natural_key_field_names` declaration. Instead of:
```

Comment:
```suggestion
Similarly, if you have a model whose natural keys include a `ForeignKey` to a model like `Location` with self-referential and variadic natural keys, for the related model to be handled properly, you must always ensure that the related field is the **last** such field in your model's uniqueness constraint or `natural_key_field_names` declaration.

Instead of this:
```